Add docs, particularly infrastructure (#36)
spacetelescope/jdaviz: docs/dev/infrastructure.rst

* *imviz*: this application provides an image-viewer for 2d images leveraging the same machinery as the other applications.  While this application is not intended to encapsulate a complete range of astronomical imaging-based workflows, it enables quick-look style visualization of images in a way that is compatible with the rest of the jdaviz framework.
* *glapp*: while not part of `jdaviz` itself, another application of relevance is the fully-flexible glue-jupyter application, designed to reproduce the functionality of the `QtGlue application <https://github.com/glue-viz/glue>`_ using this framework. While the above applications are intentionally limited in scope to focus on specific science use cases, the glapp is more flexible, so it will be desirable to have mechanisms to transition the other applications to the more complex but flexible view that the glapp can provide for the users that desire this.

Application Engine
------------------

> The applications are driven by a shared layer that connects the "high-level" layers to the "low-level" layers discussed below.  The application engine manages this.  Or more concretely, the application engine is the Python-level object that can be "shown" in any of the interfaces to actually interact with a particular application.  It contains several sub-pieces to acheive this goal.  The most directly-used portion of this is the "layout configuration" management - `jdaviz` applications specify their UI layout they use via this part of the application engine. The application engine then actually performs the layout using glue-jupyter_, `ipywidgets`, and other layout javascript/python libraries like ipyvuetify_.  The application engine also manages events and data, primarily via the `glue-core <https://github.com/glue-viz/glue>`_ library and its built-in functionality for both. In this sense the application engine also provides the interface for registering new UI functionality (although this happens indrectly via glue).

Reviewer: ```suggestion
The applications are driven by a shared layer that connects the "high-level" layers to the "low-level" layers discussed below.  The application engine manages this shared layer.  Or more concretely, the application engine is the Python-level object that can be accessed by the user in any of the interfaces to interact with a particular application.  It contains several sub-pieces to achieve this goal.  The most directly-used portion of this is the layout configuration management - `jdaviz` applications specify the UI layout they use via this part of the application engine. The application engine then constructs the layout using glue-jupyter_, `ipywidgets`, and other layout javascript/python libraries like ipyvuetify_. The application engine also leverages the built-in functionality of the `glue-core <https://github.com/glue-viz/glue>`_ library in managing application-level events and data. In this sense the application engine also provides the interface for registering new UI functionality (although this happens indirectly via glue). 
```

Some notes. I do find this _really_ vague, but that's probably unavoidable. I'm not quite sure what you were getting at in some cases, so I tried to be a bit more specified with how I interpreted your intent. I'm also not sure if the last sentence is supposed to be swapped with the second to last sentence?

Author: I hadn't thought about it but I guess it does seem better swapped now that you mention it.

Note that most of the application engine implementation belongs in glue-jupyter or glue core, as it is not unique to `jdaviz` (or even astronomy).  However, development may occur it times in `jdaviz` as a development convenience.  For example, when testing how well a web site interface works with a specific `jdaviz` pplication, it may be best implemented first in `jdaviz` using sub-classes of the glue classes, and only after the code is functional and stable is the code migrated upstream to `glue`.

Component Widgets
=================

> The "component widget" layer is the first of the "low-level" layers - i.e., the layers that actual implement specific visualization and analysis functionality. These widgets are self-contained and while they are meant to be composed in applications, for the notebook interface component widgets can and should be used directly by users for specialized scientific workflows.  Component widgets in principle can be developed in any framework that can be exposed as an `ipywidgets` widget, although currently the plan is that most will be glue-jupyter_ viewers combined with ipyvuetify_ layouts. The general goal is to push any functionality necessary for these widgets upstream into glue-jupyter_ and not confine them to `jdaviz`, but with allowances that some customization may need to be necessary for `jdaviz`/JWST-specific elements.

Reviewer: ```suggestion
The "component widget" layer is the first of the "low-level" layers - i.e., the layers that actual implement specific visualization and analysis functionality. These widgets are self-contained and in general are meant to be composed in applications. However, for the notebook interface, component widgets can and should be used directly by users for specialized scientific workflows.  Component widgets in principle can be developed in any framework that can be exposed as an `ipywidgets` widget, although currently the plan is that most will be glue-jupyter_ viewers combined with ipyvuetify_ layouts. The general goal is to push any functionality necessary for these widgets upstream into glue-jupyter_ and not confine them to `jdaviz`, but with allowances that some customization may need to be necessary for `jdaviz`/JWST-specific elements.
```
